projects/eslint-config/plugins/liferay/lib/rules/no-abbreviations.js

```diff
@@ -0,0 +1,67 @@
+/**
+ * SPDX-FileCopyrightText: © 2017 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: MIT
+ */
+
+const ABBREVIATIONS_REPLACEMENTS = {
+	arr: ['array'],
+	btn: ['button'],
+	cb: ['callback'],
+	desc: ['description'],
+	e: ['event', 'error'],
+	el: ['element'],
+	err: ['error'],
+	evt: ['event'],
+	fm: ['form'],
+	fmt: ['format'],
+	idx: ['index'],
+	img: ['image'],
+	lg: ['large'],
+	md: ['medium'],
+	obj: ['object'],
+	opts: ['options'],
+	prj: ['project'],
+	sm: ['small'],
+};
+
+const formatMessage = function (abbreviation, replacements) {
+	if (replacements.length > 1) {
+		return `Avoid ${abbreviation} abbreviation, preferred alternatives are [${replacements.join(
+			','
+		)}]`;
+	}
+	else {
+		return `Avoid ${abbreviation} abbreviation, preferred alternative is ${replacements}`;
+	}
+};
+
+module.exports = {
+	create(context) {
+		return {
+			Identifier(node) {
+				if (ABBREVIATIONS_REPLACEMENTS[node.name]) {
+					context.report({
+						message: formatMessage(
+							node.name,
+							ABBREVIATIONS_REPLACEMENTS[node.name]
+						),
+						node,
+					});
+				}
+			},
+		};
+	},
+
+	meta: {
+		docs: {
+			category: 'Best Practices',
+			description: 'Avoid abbreviations',
+			recommended: false,
+			url:
+				'https://github.com/liferay/liferay-frontend-projects/blob/master/guidelines/general/naming.md#avoid-abbreviations',
+		},
+		fixable: 'code',
```

Comment:
Since we aren't providing the automatic fix for them, we can set this as `null`

Reply:
Well, I think it is good if we provide a fix for the simplest cases (which are the ones that have only one possible replacement - most of them-), and I did it in this [commit](https://github.com/liferay/liferay-frontend-projects/pull/564/commits/bf9a821bf1bf3aaa65802031489d731bf96cec68). Maybe before merging we should fix all the abbreviations in the current project... or... would be better if we avoid this?

Reply:
I'm a bit worried about that fix :thinking: I usually assume that SF and `lint --fix` are safe operations that can be blindly trusted, but this rename of identifiers may break things IINM. For example, this code:

```javascript
const x = new YourPreferredAPI();

x.cb = 1;
```

would be transformed to:

```javascript
const x = new YourPreferredAPI();

x.callback = 1;
```

Reply:
Seeing that :point_up: I'm realizing that the fix should be OK as long assuming the detection of the problem is safe. So maybe, it's not the fix what needs to be changed, but the logic checking the problem.

In other words, it doesn't seem enough to check for `Identifier` only, but we need to make sure that the `Identifier` is referring to a local variable or parameter, for example :thinking: 

Babel has tools to check the scope of Identifiers, I assume that eslint can do the same...

Reply:
Good catch @izaera, you are rigth, that example you wrote would be transformed by the current code. I agree with you that is in the detection phase where that `.cb` should not be detected as a lint error. I will work on this 👍

Reply:
Yeah, I suffered it quite a lot of times when coding the bundler :sweat_smile:

Reply:
So... I spent all the morning trying to find a solution for this case but I couldn't... I don't see any way to distinguish the `Identifier` type that fits in this particular case... 

Anyway, I was thinking about it and maybe is it not an error... in the example, `x.cb = 1`; will be catched and fixed but also the `cb` declaration. The only thing that could cause errors is when `YourPreferredAPI` is an external library... but it is not common that other libraries exposes methods or attributes with these abbreviations that we are trying to catch... but I know, I'm assuming things. 

In addition, I found found [project](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/rules/prevent-abbreviations.js) doing something similar, and this kind of code [`this.err = 1`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/test/prevent-abbreviations.mjs#L318) is also catched as an error... 

Or another solution could be not to provide a fix (but use it before to fix oour current linting errors), forcing people to manually fix them....

Reply:
IMO, I think for fixable rules we should distinguish between rewriting vs reorganizing someones code. I would prefer a user updates their own code if it is rewriting. Reorganizing, such as sort/order, we can be more confident in making those changes with autofix. I think the potential bugs with autofixing abbreviations outweigh the gain of having them autofix. They aren't hard changes to make manually and users will learn the pattern overtime.

So I would probably not worry about making this rule automatically fixable

Reply:
> They aren't hard changes to make manually and users will learn the pattern overtime.

This applies to new code after the rule is added. Fixing 1k+ warnings isn't super easy lol

Reply:
Also, I will just add, if the fix seems to work well and we don't have any edge cases, I am fine to add it. My concern was mainly having to maintain something complex or something that has a lot of edge cases where we could potentially introduce regressions.

But if it seems to work well, then let's go for it.

Reply:
[This](https://github.com/liferay/liferay-frontend-projects/blob/1cf26ad3a5a2750a5d5d64d9671e126ad02de177/maintenance/projects/js-toolkit/packages/babel-plugin-namespace-amd-define/src/index.js#L42) is how you check scopes in Babel.

However, I don't know if `eslint` provides that `scope` field or not, because it is not part of standard ASTs, so it may be a Babel-only thing...

Regarding the rest, I agree that I would prefer not having an automatic fix for this (the truth is there aren't many fixes available in general, so why taking the risk...). For the `maintenance` folder, we already need to disable a lot of warning for Senna, so we can do the same with this, I guess.

Reply:
I think [this is how you check scopes in eslint](https://github.com/liferay/liferay-frontend-projects/pull/575/files#diff-0e30dffdb2f9f531b511c9f9998453ee58834d5077ea7dc78701839c16640672R14)...